src/rust/engine/options/src/lib.rs

```diff
@@ -706,6 +709,28 @@ impl OptionParser {
     pub fn get_unconsumed_flags(&self) -> HashMap<Scope, Vec<String>> {
         self.args_tracker.get_unconsumed_flags()
     }
+
+    // Given a map from section name to valid keys for that section,
+    // returns a vec of validation error messages.
+    pub fn validate_config(
+        &self,
+        section_to_valid_keys: &HashMap<String, HashSet<String>>,
+    ) -> Vec<String> {
+        let mut errors = vec![];
+        for (source_type, source) in self.sources.iter() {
+            if let Source::Config { ordinal: _, path } = source_type {
+                if let Some(config_reader) = source.as_any().downcast_ref::<ConfigReader>() {
+                    errors.extend(
+                        config_reader
+                            .validate(section_to_valid_keys)
```

Comment:
In my mind there's an alternative to downcasting to a particular type: having a trait method with appropriate name and signature that all `OptionSource`s implement, and `ConfigReader`'s does this, while others do nothing (potentially even have it be defaulted to nothing).

Did you consider that option?

Reply:
I did consider it, but what I got stuck on was the "appropriate name and signature" part. The validation is specifically for config files. For example, we pass in the allowed keys in each section. That would have no relevance to args or env vars. And initially the return value was also a map of config file path to invalid keys in each section in that file.

However now that the return value is simplified and is just a list of error messages, it might be possible to come up with a suitable generic API. But the argument would still be funky.  I'll think on it.

Reply:
Yeah, thought on it, and preferred on balance to downcast and just act on config, rather than create a weird impedance mismatch by trying to generalize to all OptionSources.

Reply:
Fair enough. Just justifying it to myself: this is all internal code/isn't public APIs, so we know the full set of trait implementors anyway.